src/client.cpp

```diff
@@ -291,31 +281,90 @@ void CClient::OnNewConnection()
     //### TODO: END ###//
 }
 
-void CClient::OnConClientListMesReceived ( CVector<CChannelInfo> vecChanInfo )
+void CClient::OnMuteStateHasChangedReceived ( int iServerChanID, bool bIsMuted )
+{
+    // map iChanID from server channel ID to client channel ID
+    int iChanID = FindClientChannel ( iServerChanID, false );
+
+    emit MuteStateHasChangedReceived ( iChanID, bIsMuted );
+}
+
+void CClient::OnCLChannelLevelListReceived ( CHostAddress InetAddr, CVector<uint16_t> vecLevelList )
 {
-    // Upon receiving a new client list, we have to reset oldGain and newGain
-    // entries for unused channels. This ensures that a disconnected channel
-    // does not leave behind wrong cached gain values which would leak into
-    // any new channel which reused this channel id.
-    int iNumConnectedClients = vecChanInfo.Size();
+    // reorder levels from server channel order to client channel order
 
-    // Save what channel IDs are in use:
-    bool bChanIdInUse[MAX_NUM_CHANNELS] = {};
-    for ( int i = 0; i < iNumConnectedClients; i++ )
+    if ( ReorderLevelList ( vecLevelList ) )
     {
-        bChanIdInUse[vecChanInfo[i].iChanID] = true;
+        emit CLChannelLevelListReceived ( InetAddr, vecLevelList );
     }
+}
+
+void CClient::OnConClientListMesReceived ( CVector<CChannelInfo> vecChanInfo )
+{
+    // translate from server channel IDs to client channel IDs
+    // ALSO here is where we allocate and free client channels as required
 
-    // Reset all gains for unused channel IDs:
-    for ( int iId = 0; iId < MAX_NUM_CHANNELS; iId++ )
+    const int iNumConnectedClients = vecChanInfo.Size();
+    int       i, iSrvIdx;
+
+    // on a new connection, a server sends an empty channel list before sending
+    // the real channel list (see #1010). To avoid this discarding "our" channel
+    // that we have just created, we skip this processing and just pass the empty
+    // list to the emitted signal.
+
+    if ( iNumConnectedClients != 0 )
```

Comment:
I'm not keen on a long indented block like this with two loops.  It's code complexity itches :).

Can we return early?  Or refactor to two called methods (inline maybe)?

Reply:
I think it's short enough that it can still be comprehended easily.

We could do an early return, but then we'd have to duplicate `emit ConClientListMesReceived ( vecChanInfo );` before the early return, which smells to me worse than the conditional block.

Refactoring to multiple methods means having to jump about between functions to follow the code. There is already too much of that within Jamulus!

Reply:
And we don't want to expose the `if` logic to the caller.  OK, leave as is.